marimo/_output/formatters/df_formatters.py

```diff
@@ -206,3 +208,368 @@ def _show_connect_df(
+                    """
+                    is_lazy, has_unbound = _is_lazy_display(scalar_expr)
+
+                    if is_lazy:
+                        return _format_expr_and_sql(
+                            scalar_expr, has_unbound, "Scalar"
+                        )
+                    else:
+                        try:
+                            val = scalar_expr.to_pyarrow().as_py()
+                            return ("text/plain", str(val))
+                        except BaseException as e:
+                            LOGGER.error("Failed to format Ibis scalar: %s", e)
+                            try:
+                                return ("text/plain", str(scalar_expr))
+                            except Exception:
+                                return (
+                                    "text/plain",
+                                    f"<{type(scalar_expr).__name__} object - could not display>",
+                                )
+
+            for scalar_type in json_scalar_types:
+
+                @formatting.opinionated_formatter(scalar_type)
+                def _show_marimo_ibis_json_scalar(
+                    scalar_expr,
+                ) -> tuple[KnownMimeType, str]:
+                    """Format complex JSON Scalar expressions.
+
+                    Interactive+bound expressions execute to JSON widgets.
+                    All other cases show Expression+SQL tabs.
+
+                    Examples:
+                    --------
+                    >>> ibis.array([1, 2, 3])  # Literal ArrayScalar
+                    >>> t.some_array.first()  # Column-derived ArrayScalar
+                    >>> ibis.map({"a": 1, "b": 2})  # Literal MapScalar
+                    """
+                    is_lazy, has_unbound = _is_lazy_display(scalar_expr)
+
+                    if is_lazy:
+                        # Show scalar as expression + SQL tabs
+                        return _format_expr_and_sql(
+                            scalar_expr, has_unbound, "JSON Scalar"
+                        )
+                    else:
+                        # Interactive mode with bound expression - try to execute
+                        try:
+                            val = scalar_expr.to_pyarrow().as_py()
+                            return json_output(json_data=val)._mime_()
+                        except BaseException as e:
+                            LOGGER.error("Failed to format Ibis scalar: %s", e)
+                            # Simple fallback - just show the scalar as text
+                            try:
+                                return ("text/plain", str(scalar_expr))
+                            except Exception:
+                                return (
+                                    "text/plain",
+                                    f"<{type(scalar_expr).__name__} object - could not display>",
+                                )
+
+        register_scalar_formatters()
```

Comment:
would it be possible to write some tests for these formatters? you can pattern match off of `tests/_output/formatters/test_pandas.py`